packages/eslint-plugin/src/rules/promise-function-async.ts

```diff
@@ -149,7 +134,16 @@ export default createRule<Options, MessageIds>({
         return;
       }
 
-      if (isTypeFlagSet(returnType, ts.TypeFlags.Any | ts.TypeFlags.Unknown)) {
+      const returnTypes = signatures.map(signature =>
+        checker.getReturnTypeOfSignature(signature),
+      );
+
+      if (
+        !allowAny &&
```

Comment:
This is currently not needed, and is essentially checked by `containsAllTypesByName` (https://github.com/typescript-eslint/typescript-eslint/blob/6ce04d548efc3fd9888ce731003e71c31e095d0e/packages/eslint-plugin/src/rules/promise-function-async.ts#L128). It is necessary to add this, along with checking the order of checks, so that if `allowAny` is false and a single return type is `any` or `unknown`, then the function is flagged.